dali/pipeline/operators/arithmetic/expression_tree.h

```diff
@@ -0,0 +1,174 @@
+  int64_t extent_size;
+};
+
+inline std::ostream& operator<<(std::ostream& os, const SampleExtent& v) {
+  os << "{ " << v.sample_idx << ", " << v.extent_idx << ", " << v.task_idx << ", " << v.extent_size
+     << "}";
+  return os;
+}
+
+/**
+ * @brief Part responsible for loop-execution over tile, the implementation knows the types
+ * and can access the inputs internally
+ */
+class ExpressionImplBase {
+ public:
+  virtual void Execute(ArgumentWorkspace &ws, const OpSpec &spec, SampleExtent tile) = 0;
+  virtual ~ExpressionImplBase() = default;
+};
+
+/**
+ * @brief base class for nodes of ExpressionTree
+ *
+ * TODO(klecki): Using unique_ptrs for ownership, we may generalize it to DAG, that would require
+ * some changes
+ */
+struct ExpressionBase {
+
+  ExpressionBase() = default;
+
+  virtual const std::string &GetOp() const {
+    DALI_FAIL("No op in this expression node");
+  }
+  virtual bool IsLeaf() const {
+    return false;
+  }
+  virtual bool IsTensor() const {
+    return false;
+  }
+  virtual int GetMappedInput() const {
+    return -1;
+  }
+
+  DALIDataType type_id = DALI_NO_TYPE;
+  kernels::TensorListShape<> shape;
+  std::unique_ptr<ExpressionImplBase> impl;
+
+  virtual int GetSubexpressionCount() const {
+    return 0;
+  }
+
+  virtual void AddSubexpression(std::unique_ptr<ExpressionBase> expr) {
+    DALI_FAIL("No subexpressions for this node");
+  }
+
+  virtual void SetSubexpression(int i, std::unique_ptr<ExpressionBase> expr) {
+    DALI_FAIL("No subexpressions for this node");
+  }
+
+  virtual ExpressionBase &GetSubexpression(int i) const {
+    DALI_FAIL("No subexpressions for this node");
+  }
+
+  virtual ExpressionBase &operator[](int i) {
+    DALI_FAIL("No subexpressions for this node");
+  }
+
+  virtual const ExpressionBase &operator[](int i) const {
+    DALI_FAIL("No subexpressions for this node");
+  }
+
+  virtual ~ExpressionBase() = default;
+};
+
+
+class ExpressionNode : public ExpressionBase {
+ public:
+  explicit ExpressionNode(const std::string &op) : op_(op) {
+  }
+
+  ExpressionNode(const std::string &op, int num_input) : op_(op) {
+    subexpr_.resize(num_input);
+  }
+
+  virtual int GetSubexpressionCount() const override {
+    return subexpr_.size();
+  }
+
+  void SetSubexpression(int i, std::unique_ptr<ExpressionBase> expr) override {
+    subexpr_[i] = std::move(expr);
+  }
+
+  void AddSubexpression(std::unique_ptr<ExpressionBase> expr) override {
+    subexpr_.push_back(std::move(expr));
+  }
+
+  ExpressionBase &GetSubexpression(int i) const override {
+    return *subexpr_[i];
+  }
+
+  ExpressionBase &operator[](int i) override {
+    return *subexpr_[i];
+  }
+
+  const ExpressionBase &operator[](int i) const override {
+    return *subexpr_[i];
+  }
+
+  const std::string &GetOp() const override {
+    return op_;
+  }
+
+ private:
+  std::string op_;
+  std::vector<std::unique_ptr<ExpressionBase>> subexpr_;
+};
+
+class ExpressionLeaf : public ExpressionBase {
```

Comment:
Why not have a class for scalar and tensor node? Also, I'd prefer the name "constant" over "scalar". We're going to have run-time scalars taken from Nx1 tensor(list)s.

Reply:
Done